src/track/beatmap.cpp

```diff
@@ -566,46 +572,46 @@ void BeatMap::scale(enum BPMScale scale) {
 }
 
 void BeatMap::scaleDouble() {
-    Beat prevBeat = m_beats.first();
+    BeatPointer prevBeat = m_beats.first();
     // Skip the first beat to preserve the first beat in a measure
     BeatList::iterator it = m_beats.begin() + 1;
     for (; it != m_beats.end(); ++it) {
         // Need to not accrue fractional frames.
-        int distance = it->frame_position() - prevBeat.frame_position();
-        Beat beat;
-        beat.set_frame_position(prevBeat.frame_position() + distance / 2);
+        int distance = it->get()->frame_position() - prevBeat->frame_position();
+        BeatPointer beat = BeatPointer(new Beat());
+        beat->set_frame_position(prevBeat->frame_position() + distance / 2);
         it = m_beats.insert(it, beat);
         prevBeat = (++it)[0];
     }
 }
 
 void BeatMap::scaleTriple() {
-    Beat prevBeat = m_beats.first();
+    BeatPointer prevBeat = m_beats.first();
     // Skip the first beat to preserve the first beat in a measure
     BeatList::iterator it = m_beats.begin() + 1;
     for (; it != m_beats.end(); ++it) {
         // Need to not accrue fractional frames.
-        int distance = it->frame_position() - prevBeat.frame_position();
-        Beat beat;
-        beat.set_frame_position(prevBeat.frame_position() + distance / 3);
+        int distance = it->get()->frame_position() - prevBeat->frame_position();
+        BeatPointer beat = BeatPointer(new Beat());
+        beat->set_frame_position(prevBeat->frame_position() + distance / 3);
         it = m_beats.insert(it, beat);
         ++it;
-        beat.set_frame_position(prevBeat.frame_position() + distance * 2 / 3);
+        beat->set_frame_position(prevBeat->frame_position() + distance * 2 / 3);
         it = m_beats.insert(it, beat);
         prevBeat = (++it)[0];
     }
 }
 
 void BeatMap::scaleQuadruple() {
-    Beat prevBeat = m_beats.first();
+    BeatPointer prevBeat = m_beats.first();
     // Skip the first beat to preserve the first beat in a measure
     BeatList::iterator it = m_beats.begin() + 1;
     for (; it != m_beats.end(); ++it) {
         // Need to not accrue fractional frames.
-        int distance = it->frame_position() - prevBeat.frame_position();
-        Beat beat;
+        int distance = it->get()->frame_position() - prevBeat->frame_position();
+        BeatPointer beat = BeatPointer(new Beat());
         for (int i = 1; i <= 3; i++) {
```

Comment:
Both literals 3 and 4 need to be replaced in terms of kDefaultBeatsPerMeasure. Please check all the code for these literals and implicit assumptions.

Reply:
This is quadruple scaling, which means we insert three beats between two adjacent beats. It does not depend on Time Signature. However, downbeat adjustment will make use of `kDefaultBeatsPerMeasure` while scaling.